### ISSUE TITLE
dojox.mobile addChild doesn't have second "insertIndex" parameter

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#734]
@peller, @billreed63, @jhpedemonte

Per bug #728, if you change Widget-Specific properties on an IconItem that is a child of an IconContainer, the IconItems gets reordered, with the altered IconItem becoming the last child of the IconContainer. 

Digging deeper into what's going on, there are some serious widget management things happening with the dojox.mobile.Icon\* widgets. If you look at the DOM tree after dropping a dojox.mobile.IconContainer, you'll see this:

```
<ul ... dojotype="dojox.mobile.IconContainer" ...>
  <li class="mblIconItem"...>...</li>
  <li class="mblIconItem"...>...</li>
  <li class="mblIconItemTerminator">&nbsp;</li>
  <li class="mblIconItemSub"...>
    <h2 class="mblIconContentHeading" dojoattachpoint="closeNode">...</h2>
    <div class="mblContent" dojoattachpoint="containerNode"></div>
  </li>
  <li class="mblIconItemSub"...>
    <h2 class="mblIconContentHeading" dojoattachpoint="closeNode">...</h2>
    <div class="mblContent" dojoattachpoint="containerNode"></div>
  </li>
</ul>
```

and then looking at the source code (IconContainer.js), the `addChild()` method on the widget attempts to manage the above hierarchy of elements, and is programmed to always insert a new child at the end:

```
        addChild: function(widget){
            this.domNode.insertBefore(widget.domNode, this._terminator);
            widget.transition = this.transition;
            if(this.transition === "below"){
                this.domNode.appendChild(widget.subNode);
            }
            widget.inheritParams();
            widget._setIconAttr(widget.icon);
        }
```

Note that addChild() in IconContainer.js doesn't have a second "insertIndex" parameter like the addChild() method in _Container.js.

Seems to me that we need to ask the dojox.mobile team to implement the "insertIndex" parameter, and if we can't get instant turnaround, maybe we need to remove support for dojox.mobile.*Icon.

I grep'd for addChild in dojox.mobile. It's likely more than just IconContainer that is problematic:

```
$ grep -rn addChild * | grep -v \.html
Carousel.js:110:                this.addChild(w);
IconContainer.js:52:                view.addChild(heading);
IconContainer.js:76:        addChild: function(widget){
ScrollableView.js:50:       addChild: function(widget){
SpinWheel.js:21:                this.addChild(this.slots[i]);
ViewController.js:178:                  if(parent && parent.addChild){
ViewController.js:179:                      parent.addChild(widget);
_DataListMixin.js:81:               this.addChild(this.createListItem(item));
_DataListMixin.js:95:           this.addChild(this.createListItem(newItem));
_compat.js:226:             dojox.mobile.RoundRectList._addChild = dojox.mobile.RoundRectList.prototype.addChild;
_compat.js:242:                 addChild: function(widget){
_compat.js:243:                     dojox.mobile.RoundRectList._addChild.apply(this, arguments);
_compat.js:283:             dojox.mobile.IconContainer._addChild = dojox.mobile.IconContainer.prototype.addChild;
_compat.js:285:                 addChild: function(widget){
_compat.js:286:                     dojox.mobile.IconContainer._addChild.apply(this, arguments);
```

**Comment by jhpedemonte** [comment]
So this is caused by the fact that, in Maqetta, when changing a property, we **recreate the widget entirely**, right?  Is there a reason for that?  Should we just attempt to modify the prop on the existing widget, rather than recreating anew?

**Comment by JonFerraiolo** [comment]
It depends on the command. For some commands, we have figured out how modify a widget in place, but the ultimate fallback is to recreate the widget. In some situations, we probably could figure out how to modify a widget in place, but haven't written the code yet. But we've been through this discussion multiple times in the past, and have proven to ourselves that there will always be operations where the only viable implementation approach is to remove the old widget and add a new widget back into the document.

**Comment by peller** [comment]
filed ticket http://bugs.dojotoolkit.org/ticket/13470

**Comment by JonFerraiolo** [comment]
Any chance of direct discussion with the engineers to determine when such a fix might be available? If a fix isn't imminent, then for #728, maybe with each change to an IconItem we will need to delete/recreate the containing IconContainer.

**Comment by peller** [comment]
8a0b969e56d0477061fb1079da459a4a034a277a brought in ykami's latest changes from dojox.mobile to implement add/removeChild on IconContainer and ScrollableContainer
